Ensure visually contiguous boundaries when extending selection using Shift+Arrow keys

whsieh opened the pull request:
#### 965f70e7b6c9145e2f9977f3087f42f7aa09a1ae
<pre>
Ensure visually contiguous boundaries when extending selection using Shift+Arrow keys
<a href="https://bugs.webkit.org/show_bug.cgi?id=284981">https://bugs.webkit.org/show_bug.cgi?id=284981</a>
<a href="https://rdar.apple.com/141786406">rdar://141786406</a>

Reviewed by Aditya Keerthi and Ryosuke Niwa.

Add support for snapping to the nearest bidi text boundary that results in a visually contiguous
selection on iOS, in the case where `VisuallyContiguousBidiTextSelectionEnabled` is turned on. To
achieve this, we refactor the existing utilities in `Editing.cpp` that are currently used to adjust
a selection to nearest visually-contiguous bidi text boundaries, and expose a new helper function
that&apos;s used in `FrameSelection` when performing user-triggered selection extension.

See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-extending-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-extending.html: Added.

Add a layout test to exercise this change by extending the selection by word granularity, through
LTR and RTL paragraphs containing bidi text.

* Source/WebCore/editing/Editing.cpp:
(WebCore::findBidiBoundary):

Add a `SelectionExtentMovement` enum argument to control whether we should pick the left or right
bidi boundary when determining the adjusted boundary point that maintains visual contiguity.

(WebCore::boxWithMinimumBidiLevelBetween):

Add a helper method to find the inline box with the minimum bidi level between two
`RenderedPosition`s, and use it in a couple places below.

(WebCore::primaryDirectionForSingleLineRange):
(WebCore::makeVisuallyContiguousIfNeeded):

Pull the existing logic in `adjustToVisuallyContiguousRange` into a static helper, that returns
a visually contiguous range for the given range (or `nullopt` if there&apos;s no adjustment needed).

(WebCore::adjustToVisuallyContiguousRange):

Reimplement this in terms of `makeVisuallyContiguousIfNeeded`.

(WebCore::adjustVisibleExtentPreservingVisualContiguity):

Expose a new helper function that takes a base and (mutable) extent, and adjusts the extent in order
to produce a visually contiguous selection across bidi text.

(WebCore::visuallyClosestBidiBoundary): Deleted.
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::adjustSelectionExtentIfNeeded):

In addition to adjusting the position for `user-select: all;`, we now also ensure that the new
selection extent will result in a visually contiguous selection using the above helper.

(WebCore::FrameSelection::modifyExtendingRight):
(WebCore::FrameSelection::modifyExtendingForward):
(WebCore::FrameSelection::modifyExtendingLeft):
(WebCore::FrameSelection::modifyExtendingBackward):
(WebCore::FrameSelection::modify):

Plumb the `UserTriggered` flag into `modifyExtending*`.

(WebCore::FrameSelection::updateAppearance):
(WebCore::adjustPositionForUserSelectAll): Deleted.

Rename this to `adjustSelectionExtentIfNeeded`, and make it a private method on `FrameSelection`.

* Source/WebCore/editing/FrameSelection.h:
(WebCore::RenderedPosition::boundaryPoint const):
* Source/WebCore/editing/RenderedPosition.h:

Add a helper to make a `BoundaryPoint` from the `RenderedPosition`.

Canonical link: <a href="https://commits.webkit.org/288183@main">https://commits.webkit.org/288183@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b845d29591bc94e667d6974615e12b7ab2f74496

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82057 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1584 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36014 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86614 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33089 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84163 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9410 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63966 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21688 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74658 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44250 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1104 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28836 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31509 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29447 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88047 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9299 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72336 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70477 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71558 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17849 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15683 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14594 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9250 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14792 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9090 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12616 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10898 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1224,6 +1224,11 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
 }
 
+enum class RangeEndpointsToAdjust : uint8_t {
+    Start    = 1 << 0,
```
We typically don't align = like this.

whsieh replied:
Fixed!

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1356,55 +1371,119 @@ TextDirection primaryDirectionForSingleLineRange(const Position& start, const Po
+    auto directionAtStart = renderedStart.box()->direction();
+    auto directionAtEnd = renderedEnd.box()->direction();
     if (inSameLine(start, end)) {
         forEachRenderedBoxBetween(renderedStart, renderedEnd, [&](auto box) {
```
Should we add a new method or a helper function which builds the lowest bidi-level between two RenderedPositions?

whsieh replied:
Sounds good — created a helper function here:

```
static InlineIterator::LeafBoxIterator boxWithMinimumBidiLevelBetween(const RenderedPosition& start, const RenderedPosition& end)
{
    InlineIterator::LeafBoxIterator result;
    forEachRenderedBoxBetween(start, end, [&](auto box) {
        if (!result || box->bidiLevel() < result->bidiLevel())
            result = box;
    });
    return result;
}
```

...and used it in `primaryDirectionForSingleLineRange` and `makeVisuallyContiguousIfNeeded`. This also made me realize that `directionAtStart` and `directionAtEnd` are unused now, so I reverted that part.

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -1261,29 +1266,39 @@ static std::optional<unsigned> visualDistanceOnSameLine(const RenderedPosition&
-static std::optional<BoundaryPoint> visuallyClosestBidiBoundary(const RenderedPosition& position, unsigned bidiLevel)
+static std::optional<BoundaryPoint> findBidiBoundary(const RenderedPosition& position, unsigned bidiLevel, SelectionExtentMovement movement)
 {
+    auto makeBoundaryPoint = [](const RenderedPosition& position) -> std::optional<BoundaryPoint> {
```
Maybe we can add this to RenderedPosition.h?

whsieh replied:
Makes sense — created a new helper method on `RenderedPosition`, and used it here:

```
std::optional<BoundaryPoint> boundaryPoint() const;
```

whsieh committed:
Ensure visually contiguous boundaries when extending selection using Shift+Arrow keys
https://bugs.webkit.org/show_bug.cgi?id=284981
rdar://141786406

Reviewed by Aditya Keerthi and Ryosuke Niwa.

Add support for snapping to the nearest bidi text boundary that results in a visually contiguous
selection on iOS, in the case where `VisuallyContiguousBidiTextSelectionEnabled` is turned on. To
achieve this, we refactor the existing utilities in `Editing.cpp` that are currently used to adjust
a selection to nearest visually-contiguous bidi text boundaries, and expose a new helper function
that's used in `FrameSelection` when performing user-triggered selection extension.

See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-extending-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-extending.html: Added.

Add a layout test to exercise this change by extending the selection by word granularity, through
LTR and RTL paragraphs containing bidi text.

* Source/WebCore/editing/Editing.cpp:
(WebCore::findBidiBoundary):

Add a `SelectionExtentMovement` enum argument to control whether we should pick the left or right
bidi boundary when determining the adjusted boundary point that maintains visual contiguity.

(WebCore::boxWithMinimumBidiLevelBetween):

Add a helper method to find the inline box with the minimum bidi level between two
`RenderedPosition`s, and use it in a couple places below.

(WebCore::primaryDirectionForSingleLineRange):
(WebCore::makeVisuallyContiguousIfNeeded):

Pull the existing logic in `adjustToVisuallyContiguousRange` into a static helper, that returns
a visually contiguous range for the given range (or `nullopt` if there's no adjustment needed).

(WebCore::adjustToVisuallyContiguousRange):

Reimplement this in terms of `makeVisuallyContiguousIfNeeded`.

(WebCore::adjustVisibleExtentPreservingVisualContiguity):

Expose a new helper function that takes a base and (mutable) extent, and adjusts the extent in order
to produce a visually contiguous selection across bidi text.

(WebCore::visuallyClosestBidiBoundary): Deleted.
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::adjustSelectionExtentIfNeeded):

In addition to adjusting the position for `user-select: all;`, we now also ensure that the new
selection extent will result in a visually contiguous selection using the above helper.

(WebCore::FrameSelection::modifyExtendingRight):
(WebCore::FrameSelection::modifyExtendingForward):
(WebCore::FrameSelection::modifyExtendingLeft):
(WebCore::FrameSelection::modifyExtendingBackward):
(WebCore::FrameSelection::modify):

Plumb the `UserTriggered` flag into `modifyExtending*`.

(WebCore::FrameSelection::updateAppearance):
(WebCore::adjustPositionForUserSelectAll): Deleted.

Rename this to `adjustSelectionExtentIfNeeded`, and make it a private method on `FrameSelection`.

* Source/WebCore/editing/FrameSelection.h:
(WebCore::RenderedPosition::boundaryPoint const):
* Source/WebCore/editing/RenderedPosition.h:

Add a helper to make a `BoundaryPoint` from the `RenderedPosition`.

Canonical link: https://commits.webkit.org/288183@main
